perl5/perls/perl-5.12.4/man/man3/DateTime.3
.\" Automatically generated by Pod::Man 2.23 (Pod::Simple 3.14)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.ie \nF \{\
.    de IX
.    tm Index:\\$1\t\\n%\t"\\$2"
..
.    nr % 0
.    rr F
.\}
.el \{\
.    de IX
..
.\}
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "DateTime 3"
.TH DateTime 3 "2012-02-01" "perl v5.12.4" "User Contributed Perl Documentation"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
DateTime \- A date and time object
.SH "VERSION"
.IX Header "VERSION"
version 0.72
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 1
\&  use DateTime;
\&
\&  $dt = DateTime\->new(
\&      year       => 1964,
\&      month      => 10,
\&      day        => 16,
\&      hour       => 16,
\&      minute     => 12,
\&      second     => 47,
\&      nanosecond => 500000000,
\&      time_zone  => \*(AqAsia/Taipei\*(Aq,
\&  );
\&
\&  $dt = DateTime\->from_epoch( epoch => $epoch );
\&  $dt = DateTime\->now; # same as ( epoch => time() )
\&
\&  $year   = $dt\->year;
\&  $month  = $dt\->month;          # 1\-12
\&
\&  $day    = $dt\->day;            # 1\-31
\&
\&  $dow    = $dt\->day_of_week;    # 1\-7 (Monday is 1)
\&
\&  $hour   = $dt\->hour;           # 0\-23
\&  $minute = $dt\->minute;         # 0\-59
\&
\&  $second = $dt\->second;         # 0\-61 (leap seconds!)
\&
\&  $doy    = $dt\->day_of_year;    # 1\-366 (leap years)
\&
\&  $doq    = $dt\->day_of_quarter; # 1..
\&
\&  $qtr    = $dt\->quarter;        # 1\-4
\&
\&  # all of the start\-at\-1 methods above have corresponding start\-at\-0
\&  # methods, such as $dt\->day_of_month_0, $dt\->month_0 and so on
\&
\&  $ymd    = $dt\->ymd;           # 2002\-12\-06
\&  $ymd    = $dt\->ymd(\*(Aq/\*(Aq);      # 2002/12/06
\&
\&  $mdy    = $dt\->mdy;           # 12\-06\-2002
\&  $mdy    = $dt\->mdy(\*(Aq/\*(Aq);      # 12/06/2002
\&
\&  $dmy    = $dt\->dmy;           # 06\-12\-2002
\&  $dmy    = $dt\->dmy(\*(Aq/\*(Aq);      # 06/12/2002
\&
\&  $hms    = $dt\->hms;           # 14:02:29
\&  $hms    = $dt\->hms(\*(Aq!\*(Aq);      # 14!02!29
\&
\&  $is_leap  = $dt\->is_leap_year;
\&
\&  # these are localizable, see Locales section
\&  $month_name  = $dt\->month_name; # January, February, ...
\&  $month_abbr  = $dt\->month_abbr; # Jan, Feb, ...
\&  $day_name    = $dt\->day_name;   # Monday, Tuesday, ...
\&  $day_abbr    = $dt\->day_abbr;   # Mon, Tue, ...
\&
\&  # May not work for all possible datetime, see the docs on this
\&  # method for more details.
\&  $epoch_time  = $dt\->epoch;
\&
\&  $dt2 = $dt + $duration_object;
\&
\&  $dt3 = $dt \- $duration_object;
\&
\&  $duration_object = $dt \- $dt2;
\&
\&  $dt\->set( year => 1882 );
\&
\&  $dt\->set_time_zone( \*(AqAmerica/Chicago\*(Aq );
\&
\&  $dt\->set_formatter( $formatter );
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
DateTime is a class for the representation of date/time combinations,
and is part of the Perl DateTime project. For details on this project
please see <http://datetime.perl.org/>. The DateTime site has a \s-1FAQ\s0
which may help answer many \*(L"how do I do X?\*(R" questions. The \s-1FAQ\s0 is at
<http://datetime.perl.org/wiki/datetime/page/FAQ>.
.PP
It represents the Gregorian calendar, extended backwards in time
before its creation (in 1582). This is sometimes known as the
\&\*(L"proleptic Gregorian calendar\*(R". In this calendar, the first day of
the calendar (the epoch), is the first day of year 1, which
corresponds to the date which was (incorrectly) believed to be the
birth of Jesus Christ.
.PP
The calendar represented does have a year 0, and in that way differs
from how dates are often written using \*(L"\s-1BCE/CE\s0\*(R" or \*(L"\s-1BC/AD\s0\*(R".
.PP
For infinite datetimes, please see the
DateTime::Infinite module.
.SH "USAGE"
.IX Header "USAGE"
.SS "0\-based Versus 1\-based Numbers"
.IX Subsection "0-based Versus 1-based Numbers"
The DateTime.pm module follows a simple consistent logic for
determining whether or not a given number is 0\-based or 1\-based.
.PP
Month, day of month, day of week, and day of year are 1\-based. Any
method that is 1\-based also has an equivalent 0\-based method ending in
\&\*(L"_0\*(R". So for example, this class provides both \f(CW\*(C`day_of_week()\*(C'\fR and
\&\f(CW\*(C`day_of_week_0()\*(C'\fR methods.
.PP
The \f(CW\*(C`day_of_week_0()\*(C'\fR method still treats Monday as the first day of
the week.
.PP
All \fItime\fR\-related numbers such as hour, minute, and second are
0\-based.
.PP
Years are neither, as they can be both positive or negative, unlike
any other datetime component. There \fIis\fR a year 0.
.PP
There is no \f(CW\*(C`quarter_0()\*(C'\fR method.
.SS "Error Handling"
.IX Subsection "Error Handling"
Some errors may cause this module to die with an error string. This
can only happen when calling constructor methods, methods that change
the object, such as \f(CW\*(C`set()\*(C'\fR, or methods that take parameters.
Methods that retrieve information about the object, such as \f(CW\*(C`year()\*(C'\fR
or \f(CW\*(C`epoch()\*(C'\fR, will never die.
.SS "Locales"
.IX Subsection "Locales"
All the object methods which return names or abbreviations return data
based on a locale. This is done by setting the locale when
constructing a DateTime object. There is also a \f(CW\*(C`DefaultLocale()\*(C'\fR
class method which may be used to set the default locale for all
DateTime objects created. If this is not set, then \*(L"en_US\*(R" is used.
.SS "Floating DateTimes"
.IX Subsection "Floating DateTimes"
The default time zone for new DateTime objects, except where stated
otherwise, is the \*(L"floating\*(R" time zone. This concept comes from the
iCal standard. A floating datetime is one which is not anchored to
any particular time zone. In addition, floating datetimes do not
include leap seconds, since we cannot apply them without knowing the
datetime's time zone.
.PP
The results of date math and comparison between a floating datetime
and one with a real time zone are not really valid, because one
includes leap seconds and the other does not. Similarly, the results
of datetime math between two floating datetimes and two datetimes with
time zones are not really comparable.
.PP
If you are planning to use any objects with a real time zone, it is
strongly recommended that you \fBdo not\fR mix these with floating
datetimes.
.SS "Math"
.IX Subsection "Math"
If you are going to be using doing date math, please read the section \*(L"How
Datetime Math Works\*(R".
.SS "Time Zone Warnings"
.IX Subsection "Time Zone Warnings"
Determining the local time zone for a system can be slow. If \f(CW$ENV{TZ}\fR is
not set, it may involve reading a number of files in \fI/etc\fR or elsewhere. If
you know that the local time zone won't change while your code is running, and
you need to make many objects for the local time zone, it is strongly
recommended that you retrieve the local time zone once and cache it:
.PP
.Vb 1
\&  our $App::LocalTZ = DateTime::TimeZone\->new( name => \*(Aqlocal\*(Aq );
\&
\&  ... # then everywhere else
\&
\&  my $dt = DateTime\->new( ..., time_zone => $App::LocalTZ );
.Ve
.PP
DateTime itself does not do this internally because local time zones can
change, and there's no good way to determine if it's changed without doing all
the work to look it up.
.PP
Do not try to use named time zones (like \*(L"America/Chicago\*(R") with dates
very far in the future (thousands of years). The current
implementation of \f(CW\*(C`DateTime::TimeZone\*(C'\fR will use a huge amount of
memory calculating all the \s-1DST\s0 changes from now until the future
date. Use \s-1UTC\s0 or the floating time zone and you will be safe.
.SS "Methods"
.IX Subsection "Methods"
\fIConstructors\fR
.IX Subsection "Constructors"
.PP
All constructors can die when invalid parameters are given.
.IP "\(bu" 4
DateTime\->new( ... )
.Sp
This class method accepts parameters for each date and time component:
\&\*(L"year\*(R", \*(L"month\*(R", \*(L"day\*(R", \*(L"hour\*(R", \*(L"minute\*(R", \*(L"second\*(R", \*(L"nanosecond\*(R".
It also accepts \*(L"locale\*(R", \*(L"time_zone\*(R", and \*(L"formatter\*(R" parameters.
.Sp
.Vb 10
\&  my $dt = DateTime\->new(
\&      year       => 1966,
\&      month      => 10,
\&      day        => 25,
\&      hour       => 7,
\&      minute     => 15,
\&      second     => 47,
\&      nanosecond => 500000000,
\&      time_zone  => \*(AqAmerica/Chicago\*(Aq,
\&  );
.Ve
.Sp
DateTime validates the \*(L"month\*(R", \*(L"day\*(R", \*(L"hour\*(R", \*(L"minute\*(R", and \*(L"second\*(R",
and \*(L"nanosecond\*(R" parameters. The valid values for these parameters are:
.RS 4
.IP "\(bu" 8
month
.Sp
An integer from 1\-12.
.IP "\(bu" 8
day
.Sp
An integer from 1\-31, and it must be within the valid range of days for the
specified month.
.IP "\(bu" 8
hour
.Sp
An integer from 0\-23.
.IP "\(bu" 8
minute
.Sp
An integer from 0\-59.
.IP "\(bu" 8
second
.Sp
An integer from 0\-61 (to allow for leap seconds). Values of 60 or 61 are only
allowed when they match actual leap seconds.
.IP "\(bu" 8
nanosecond
.Sp
An integer >= 0. If this number is greater than 1 billion, it will be
normalized into the second value for the DateTime object.
.RE
.RS 4
.Sp
Invalid parameter types (like an array reference) will cause the
constructor to die.
.Sp
The value for seconds may be from 0 to 61, to account for leap
seconds. If you give a value greater than 59, DateTime does check to
see that it really matches a valid leap second.
.Sp
All of the parameters are optional except for \*(L"year\*(R". The \*(L"month\*(R" and
\&\*(L"day\*(R" parameters both default to 1, while the \*(L"hour\*(R", \*(L"minute\*(R",
\&\*(L"second\*(R", and \*(L"nanosecond\*(R" parameters all default to 0.
.Sp
The \*(L"locale\*(R" parameter should be a string matching one of the valid
locales, or a \f(CW\*(C`DateTime::Locale\*(C'\fR object. See the
DateTime::Locale documentation for details.
.Sp
The time_zone parameter can be either a scalar or a
\&\f(CW\*(C`DateTime::TimeZone\*(C'\fR object. A string will simply be passed to the
\&\f(CW\*(C`DateTime::TimeZone\->new\*(C'\fR method as its \*(L"name\*(R" parameter. This
string may be an Olson \s-1DB\s0 time zone name (\*(L"America/Chicago\*(R"), an
offset string (\*(L"+0630\*(R"), or the words \*(L"floating\*(R" or \*(L"local\*(R". See the
\&\f(CW\*(C`DateTime::TimeZone\*(C'\fR documentation for more details.
.Sp
The default time zone is \*(L"floating\*(R".
.Sp
The \*(L"formatter\*(R" can be either a scalar or an object, but the class
specified by the scalar or the object must implement a
\&\f(CW\*(C`format_datetime()\*(C'\fR method.
.RE
.PP
Parsing Dates
.IX Subsection "Parsing Dates"
.PP
\&\fBThis module does not parse dates!\fR That means there is no
constructor to which you can pass things like \*(L"March 3, 1970 12:34\*(R".
.PP
Instead, take a look at the various \f(CW\*(C`DateTime::Format::*\*(C'\fR modules on
\&\s-1CPAN\s0. These parse all sorts of different date formats, and you're
bound to find something that can handle your particular needs.
.PP
Ambiguous Local Times
.IX Subsection "Ambiguous Local Times"
.PP
Because of Daylight Saving Time, it is possible to specify a local
time that is ambiguous. For example, in the \s-1US\s0 in 2003, the
transition from to saving to standard time occurred on October 26, at
02:00:00 local time. The local clock changed from 01:59:59 (saving
time) to 01:00:00 (standard time). This means that the hour from
01:00:00 through 01:59:59 actually occurs twice, though the \s-1UTC\s0 time
continues to move forward.
.PP
If you specify an ambiguous time, then the latest \s-1UTC\s0 time is always
used, in effect always choosing standard time. In this case, you can
simply subtract an hour to the object in order to move to saving time,
for example:
.PP
.Vb 10
\&  # This object represent 01:30:00 standard time
\&  my $dt = DateTime\->new(
\&      year      => 2003,
\&      month     => 10,
\&      day       => 26,
\&      hour      => 1,
\&      minute    => 30,
\&      second    => 0,
\&      time_zone => \*(AqAmerica/Chicago\*(Aq,
\&  );
\&
\&  print $dt\->hms;  # prints 01:30:00
\&
\&  # Now the object represent 01:30:00 saving time
\&  $dt\->subtract( hours => 1 );
\&
\&  print $dt\->hms;  # still prints 01:30:00
.Ve
.PP
Alternately, you could create the object with the \s-1UTC\s0 time zone, and
then call the \f(CW\*(C`set_time_zone()\*(C'\fR method to change the time zone. This
is a good way to ensure that the time is not ambiguous.
.PP
Invalid Local Times
.IX Subsection "Invalid Local Times"
.PP
Another problem introduced by Daylight Saving Time is that certain
local times just do not exist. For example, in the \s-1US\s0 in 2003, the
transition from standard to saving time occurred on April 6, at the
change to 2:00:00 local time. The local clock changes from 01:59:59
(standard time) to 03:00:00 (saving time). This means that there is
no 02:00:00 through 02:59:59 on April 6!
.PP
Attempting to create an invalid time currently causes a fatal error.
This may change in future version of this module.
.IP "\(bu" 4
DateTime\->from_epoch( epoch => \f(CW$epoch\fR, ... )
.Sp
This class method can be used to construct a new DateTime object from
an epoch time instead of components. Just as with the \f(CW\*(C`new()\*(C'\fR
method, it accepts \*(L"time_zone\*(R", \*(L"locale\*(R", and \*(L"formatter\*(R" parameters.
.Sp
If the epoch value is not an integer, the part after the decimal will
be converted to nanoseconds. This is done in order to be compatible
with \f(CW\*(C`Time::HiRes\*(C'\fR. If the floating portion extends past 9 decimal
places, it will be truncated to nine, so that 1.1234567891 will become
1 second and 123,456,789 nanoseconds.
.Sp
By default, the returned object will be in the \s-1UTC\s0 time zone.
.IP "\(bu" 4
DateTime\->now( ... )
.Sp
This class method is equivalent to calling \f(CW\*(C`from_epoch()\*(C'\fR with the
value returned from Perl's \f(CW\*(C`time()\*(C'\fR function. Just as with the
\&\f(CW\*(C`new()\*(C'\fR method, it accepts \*(L"time_zone\*(R" and \*(L"locale\*(R" parameters.
.Sp
By default, the returned object will be in the \s-1UTC\s0 time zone.
.IP "\(bu" 4
DateTime\->today( ... )
.Sp
This class method is equivalent to:
.Sp
.Vb 1
\&  DateTime\->now\->truncate( to => \*(Aqday\*(Aq );
.Ve
.IP "\(bu" 4
DateTime\->from_object( object => \f(CW$object\fR, ... )
.Sp
This class method can be used to construct a new DateTime object from
any object that implements the \f(CW\*(C`utc_rd_values()\*(C'\fR method. All
\&\f(CW\*(C`DateTime::Calendar\*(C'\fR modules must implement this method in order to
provide cross-calendar compatibility. This method accepts a
\&\*(L"locale\*(R" and \*(L"formatter\*(R" parameter
.Sp
If the object passed to this method has a \f(CW\*(C`time_zone()\*(C'\fR method, that
is used to set the time zone of the newly created \f(CW\*(C`DateTime.pm\*(C'\fR
object.
.Sp
Otherwise, the returned object will be in the floating time zone.
.IP "\(bu" 4
DateTime\->last_day_of_month( ... )
.Sp
This constructor takes the same arguments as can be given to the
\&\f(CW\*(C`new()\*(C'\fR method, except for \*(L"day\*(R". Additionally, both \*(L"year\*(R" and
\&\*(L"month\*(R" are required.
.IP "\(bu" 4
DateTime\->from_day_of_year( ... )
.Sp
This constructor takes the same arguments as can be given to the
\&\f(CW\*(C`new()\*(C'\fR method, except that it does not accept a \*(L"month\*(R" or \*(L"day\*(R"
argument. Instead, it requires both \*(L"year\*(R" and \*(L"day_of_year\*(R". The
day of year must be between 1 and 366, and 366 is only allowed for
leap years.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIclone()\fR
.Sp
This object method returns a new object that is replica of the object
upon which the method is called.
.PP
\fI\*(L"Get\*(R" Methods\fR
.IX Subsection "Get Methods"
.PP
This class has many methods for retrieving information about an
object.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIyear()\fR
.Sp
Returns the year.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIce_year()\fR
.Sp
Returns the year according to the \s-1BCE/CE\s0 numbering system. The year
before year 1 in this system is year \-1, aka \*(L"1 \s-1BCE\s0\*(R".
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIera_name()\fR
.Sp
Returns the long name of the current era, something like \*(L"Before
Christ\*(R". See the Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIera_abbr()\fR
.Sp
Returns the abbreviated name of the current era, something like \*(L"\s-1BC\s0\*(R".
See the Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIchristian_era()\fR
.Sp
Returns a string, either \*(L"\s-1BC\s0\*(R" or \*(L"\s-1AD\s0\*(R", according to the year.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIsecular_era()\fR
.Sp
Returns a string, either \*(L"\s-1BCE\s0\*(R" or \*(L"\s-1CE\s0\*(R", according to the year.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIyear_with_era()\fR
.Sp
Returns a string containing the year immediately followed by its era
abbreviation. The year is the absolute value of \f(CW\*(C`ce_year()\*(C'\fR, so that
year 1 is \*(L"1AD\*(R" and year 0 is \*(L"1BC\*(R".
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIyear_with_christian_era()\fR
.Sp
Like \f(CW\*(C`year_with_era()\*(C'\fR, but uses the \fIchristian_era()\fR to get the era
name.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIyear_with_secular_era()\fR
.Sp
Like \f(CW\*(C`year_with_era()\*(C'\fR, but uses the \fIsecular_era()\fR method to get the
era name.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fImonth()\fR
.Sp
Returns the month of the year, from 1..12.
.Sp
Also available as \f(CW\*(C`$dt\->mon()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fImonth_name()\fR
.Sp
Returns the name of the current month. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fImonth_abbr()\fR
.Sp
Returns the abbreviated name of the current month. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday()\fR
.Sp
Returns the day of the month, from 1..31.
.Sp
Also available as \f(CW\*(C`$dt\->mday()\*(C'\fR and \f(CW\*(C`$dt\->day_of_month()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday_of_week()\fR
.Sp
Returns the day of the week as a number, from 1..7, with 1 being
Monday and 7 being Sunday.
.Sp
Also available as \f(CW\*(C`$dt\->wday()\*(C'\fR and \f(CW\*(C`$dt\->dow()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIlocal_day_of_week()\fR
.Sp
Returns the day of the week as a number, from 1..7. The day
corresponding to 1 will vary based on the locale.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday_name()\fR
.Sp
Returns the name of the current day of the week. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday_abbr()\fR
.Sp
Returns the abbreviated name of the current day of the week. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday_of_year()\fR
.Sp
Returns the day of the year.
.Sp
Also available as \f(CW\*(C`$dt\->doy()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIquarter()\fR
.Sp
Returns the quarter of the year, from 1..4.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIquarter_name()\fR
.Sp
Returns the name of the current quarter. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIquarter_abbr()\fR
.Sp
Returns the abbreviated name of the current quarter. See the
Locales section for more details.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIday_of_quarter()\fR
.Sp
Returns the day of the quarter.
.Sp
Also available as \f(CW\*(C`$dt\->doq()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIweekday_of_month()\fR
.Sp
Returns a number from 1..5 indicating which week day of the month this
is. For example, June 9, 2003 is the second Monday of the month, and
so this method returns 2 for that day.
.IP "\(bu" 4
\&\f(CW$dt\fR\->ymd( \f(CW$optional_separator\fR )
.IP "\(bu" 4
\&\f(CW$dt\fR\->mdy( \f(CW$optional_separator\fR )
.IP "\(bu" 4
\&\f(CW$dt\fR\->dmy( \f(CW$optional_separator\fR )
.Sp
Each method returns the year, month, and day, in the order indicated
by the method name. Years are zero-padded to four digits. Months and
days are 0\-padded to two digits.
.Sp
By default, the values are separated by a dash (\-), but this can be
overridden by passing a value to the method.
.Sp
The \f(CW\*(C`$dt\->ymd()\*(C'\fR method is also available as \f(CW\*(C`$dt\->date()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIhour()\fR
.Sp
Returns the hour of the day, from 0..23.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIhour_1()\fR
.Sp
Returns the hour of the day, from 1..24.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIhour_12()\fR
.Sp
Returns the hour of the day, from 1..12.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIhour_12_0()\fR
.Sp
Returns the hour of the day, from 0..11.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIam_or_pm()\fR
.Sp
Returns the appropriate localized abbreviation, depending on the
current hour.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIminute()\fR
.Sp
Returns the minute of the hour, from 0..59.
.Sp
Also available as \f(CW\*(C`$dt\->min()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIsecond()\fR
.Sp
Returns the second, from 0..61. The values 60 and 61 are used for
leap seconds.
.Sp
Also available as \f(CW\*(C`$dt\->sec()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIfractional_second()\fR
.Sp
Returns the second, as a real number from 0.0 until 61.999999999
.Sp
The values 60 and 61 are used for leap seconds.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fImillisecond()\fR
.Sp
Returns the fractional part of the second as milliseconds (1E\-3 seconds).
.Sp
Half a second is 500 milliseconds.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fImicrosecond()\fR
.Sp
Returns the fractional part of the second as microseconds (1E\-6
seconds). This value will be rounded to an integer.
.Sp
Half a second is 500_000 microseconds. This value will be rounded to
an integer.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fInanosecond()\fR
.Sp
Returns the fractional part of the second as nanoseconds (1E\-9 seconds).
.Sp
Half a second is 500_000_000 nanoseconds.
.IP "\(bu" 4
\&\f(CW$dt\fR\->hms( \f(CW$optional_separator\fR )
.Sp
Returns the hour, minute, and second, all zero-padded to two digits.
If no separator is specified, a colon (:) is used by default.
.Sp
Also available as \f(CW\*(C`$dt\->time()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIdatetime()\fR
.Sp
This method is equivalent to:
.Sp
.Vb 1
\&  $dt\->ymd(\*(Aq\-\*(Aq) . \*(AqT\*(Aq . $dt\->hms(\*(Aq:\*(Aq)
.Ve
.Sp
Also available as \f(CW\*(C`$dt\->iso8601()\*(C'\fR.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIis_leap_year()\fR
.Sp
This method returns a true or false indicating whether or not the
datetime object is in a leap year.
.IP "\(bu" 4
\&\f(CW$dt\fR\->\fIweek()\fR
.Sp
.Vb 1
\& ($week_year, $week_number) = $dt\->week;
.Ve
.Sp
Returns information about the calendar week which contains this
datetime object. The values returned by this method are also available
separately through the week_year and week_number methods.
.Sp
The first week of the year is defined by \s-1ISO\s0 as the one which contains
the fourth day of January, which is equivalent to saying that it's the
first week to overlap the new year by at least four days.
.Sp

perl5/perls/perl-5.12.4/lib/site_perl/5.12.4/HTML/FormHandler/Field/DateTime.pm
package HTML::FormHandler::Field::DateTime;
# ABSTRACT: compound DateTime field

use Moose;
extends 'HTML::FormHandler::Field::Compound';

use DateTime;
use Try::Tiny;
our $VERSION = '0.04';


has '+widget' => ( default => 'compound' );

our $class_messages = {
    'datetime_invalid' => 'Not a valid DateTime',
};
sub get_class_messages {
    my $self = shift;
    return {
        %{ $self->next::method },
        %$class_messages,
    }
}

sub deflate {
    my ( $self, $value ) = @_;
    return $value unless ref $value eq 'DateTime';
    my %hash;
    foreach my $field ( $self->all_fields ) {
        my $meth = $field->name;
        $hash{$meth} = $value->$meth;
    }
    return \%hash;
}

sub validate {
    my ($self) = @_;
    my @dt_parms;
    foreach my $child ( $self->all_fields ) {
        next unless $child->value;
        push @dt_parms, ( $child->accessor => $child->value );
    }

    # set the value
    my $dt;
    try {
        $dt = DateTime->new(@dt_parms);
    }
    catch {
        $self->add_error( $self->get_message('datetime_invalid') );
    };
    if( $dt ) {
        $self->_set_value($dt);
    }
    else {
        $self->_set_value( {@dt_parms} );
    }
}

__PACKAGE__->meta->make_immutable;
use namespace::autoclean;
1;


__END__
=pod

=head1 NAME

HTML::FormHandler::Field::DateTime - compound DateTime field

=head1 VERSION

version 0.36002

=head1 DESCRIPTION

This is a compound field that uses modified field names for the
sub fields instead of using a separate sub-form. Widget type is 'compound'.

If you want to use drop-down select boxes for your DateTime, you
can use fields like:

    has_field 'my_date' => ( type => 'DateTime' );
    has_field 'my_date.month' => ( type => 'Month' );
    has_field 'my_date.day' => ( type => 'MonthDay' );
    has_field 'my_date.year' => ( type => 'Year' );
    has_field 'my_date.hour' => ( type => 'Hour' );
    has_field 'my_date.minute' => ( type => 'Minute' );

If you want simple input fields:

    has_field 'my_date' => ( type => 'DateTime' );
    has_field 'my_date.month' => ( type => 'Integer', range_start => 1,
         range_end => 12 );
    has_field 'my_date.day' => ( type => 'Integer', range_start => 1,
         range_end => 31 );

Customizable error: 'datetime_invalid' (default = "Not a valid DateTime")

=head1 AUTHOR

FormHandler Contributors - see HTML::FormHandler

=head1 COPYRIGHT AND LICENSE

This software is copyright (c) 2012 by Gerda Shank.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.

=cut

